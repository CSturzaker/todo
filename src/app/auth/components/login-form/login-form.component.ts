import { Component, Input, Output, EventEmitter } from '@angular/core'
import { FormGroup, FormControl } from '@angular/forms'

// models
import { Credentials } from '../../models'

@Component({
  selector: 'todo-login-form',
  templateUrl: './login-form.component.html',
  styleUrls: ['./login-form.component.scss'],
})
export class LoginFormComponent {
  @Input()
  set pending(isPending: boolean) {
    if (isPending) {
      this.loginForm.disable()
    } else {
      this.loginForm.enable()
    }
  }

  @Input() errorMessage: string | null

  @Output() submitted = new EventEmitter<Credentials>()

  loginForm: FormGroup = new FormGroup({
    username: new FormControl('Guglielmo Marconi'),
    password: new FormControl(''),
  })

  submit() {
    if (this.loginForm.valid) {
      this.submitted.emit(this.loginForm.value)
    }
  }
}
