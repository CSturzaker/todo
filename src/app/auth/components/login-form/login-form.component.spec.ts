import { ComponentFixture, TestBed } from '@angular/core/testing'
import { ReactiveFormsModule } from '@angular/forms'
import { NO_ERRORS_SCHEMA } from '@angular/core'

import { LoginFormComponent } from './login-form.component'

describe('LoginFormComponent', () => {
  let component: LoginFormComponent
  let fixture: ComponentFixture<LoginFormComponent>

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [ReactiveFormsModule],
      declarations: [LoginFormComponent],
      schemas: [NO_ERRORS_SCHEMA],
    })

    fixture = TestBed.createComponent(LoginFormComponent)
    component = fixture.componentInstance
  })

  it('should create', () => {
    fixture.detectChanges()
    expect(fixture).toMatchSnapshot()
  })

  it('should disable the form if pending', () => {
    component.pending = true

    fixture.detectChanges()

    expect(fixture).toMatchSnapshot()
  })

  it('should display an error message if provided', () => {
    component.errorMessage = 'Invalid credentials'

    fixture.detectChanges()

    expect(fixture).toMatchSnapshot()
  })

  it('should emit an event if the form is valid when submitted', () => {
    const credentials = {
      username: 'user',
      password: 'pass',
    }

    component.loginForm.setValue(credentials)

    spyOn(component.submitted, 'emit')
    component.submit()

    expect(component.submitted.emit).toHaveBeenCalledWith(credentials)
  })
})
