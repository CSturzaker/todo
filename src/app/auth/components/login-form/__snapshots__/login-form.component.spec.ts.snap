// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginFormComponent should create 1`] = `
<todo-login-form
  loginForm={[Function FormGroup]}
  submitted={[Function EventEmitter]}
>
  <mat-card>
    <mat-card-title>
      Login
    </mat-card-title>
    <mat-card-content>
      <form
        class="ng-untouched ng-pristine ng-valid"
        ng-reflect-form="[object Object]"
        novalidate=""
      >
        <p>
          <mat-form-field>
            <input
              class="ng-untouched ng-pristine ng-valid"
              formcontrolname="username"
              matinput=""
              ng-reflect-name="username"
              placeholder="Username"
              type="text"
            />
          </mat-form-field>
        </p>
        <p>
          <mat-form-field>
            <input
              class="ng-untouched ng-pristine ng-valid"
              formcontrolname="password"
              matinput=""
              ng-reflect-name="password"
              placeholder="Password"
              type="password"
            />
          </mat-form-field>
        </p>
        
        <p
          class="loginButtons"
        >
          <button
            mat-button=""
            type="submit"
          >
            Login
          </button>
        </p>
      </form>
    </mat-card-content>
  </mat-card>
</todo-login-form>
`;

exports[`LoginFormComponent should disable the form if pending 1`] = `
<todo-login-form
  loginForm={[Function FormGroup]}
  submitted={[Function EventEmitter]}
>
  <mat-card>
    <mat-card-title>
      Login
    </mat-card-title>
    <mat-card-content>
      <form
        class="ng-untouched ng-pristine"
        ng-reflect-form="[object Object]"
        novalidate=""
      >
        <p>
          <mat-form-field>
            <input
              class="ng-untouched ng-pristine"
              disabled=""
              formcontrolname="username"
              matinput=""
              ng-reflect-name="username"
              placeholder="Username"
              type="text"
            />
          </mat-form-field>
        </p>
        <p>
          <mat-form-field>
            <input
              class="ng-untouched ng-pristine"
              disabled=""
              formcontrolname="password"
              matinput=""
              ng-reflect-name="password"
              placeholder="Password"
              type="password"
            />
          </mat-form-field>
        </p>
        
        <p
          class="loginButtons"
        >
          <button
            mat-button=""
            type="submit"
          >
            Login
          </button>
        </p>
      </form>
    </mat-card-content>
  </mat-card>
</todo-login-form>
`;

exports[`LoginFormComponent should display an error message if provided 1`] = `
<todo-login-form
  errorMessage={[Function String]}
  loginForm={[Function FormGroup]}
  submitted={[Function EventEmitter]}
>
  <mat-card>
    <mat-card-title>
      Login
    </mat-card-title>
    <mat-card-content>
      <form
        class="ng-untouched ng-pristine ng-valid"
        ng-reflect-form="[object Object]"
        novalidate=""
      >
        <p>
          <mat-form-field>
            <input
              class="ng-untouched ng-pristine ng-valid"
              formcontrolname="username"
              matinput=""
              ng-reflect-name="username"
              placeholder="Username"
              type="text"
            />
          </mat-form-field>
        </p>
        <p>
          <mat-form-field>
            <input
              class="ng-untouched ng-pristine ng-valid"
              formcontrolname="password"
              matinput=""
              ng-reflect-name="password"
              placeholder="Password"
              type="password"
            />
          </mat-form-field>
        </p>
        
        <p
          class="loginError"
        >
           Invalid credentials 
        </p>
        <p
          class="loginButtons"
        >
          <button
            mat-button=""
            type="submit"
          >
            Login
          </button>
        </p>
      </form>
    </mat-card-content>
  </mat-card>
</todo-login-form>
`;
